llvm/lib/Target/AIE/AIEBaseSubtarget.cpp

```diff
@@ -381,25 +383,55 @@ class PropagateIncomingLatencies : public ScheduleDAGMutation {
           }))
         continue;
 
-      // Find the common latency for all predecessors that can be
-      // "moved" to successors.
-      SDep *MinLatencyDep = nullptr;
-      for (SDep &PredEdge : make_filter_range(SU.Preds, IsData)) {
-        if (!MinLatencyDep ||
-            PredEdge.getLatency() < MinLatencyDep->getLatency())
-          MinLatencyDep = &PredEdge;
+      // Whether to propagate latency from predecessors to successors (true),
+      // or from successors to predecessors (false).
+      bool MoveLatToSuccessors = true;
+
+      // Do not change the latency if the REG_SEQUENCE has one source
+      // outside this MBB. Such REG_SEQUENCE instructions will typically require
+      // a COPY from the external source to one of the lanes of the destination
+      // register. It is important to then keep the REG_SEQUENCE close to its
+      // consumers, because after MachinePipeliner, this typically means that
+      // only the lanes corresponding to internal sources will be loop-carried.
+      // The external lane will come directly from the pre-header, and the
+      // required COPY can then be hoisted by MachineLICM.
+      const MachineBasicBlock &MBB = *MI.getParent();
+      const MachineRegisterInfo &MRI = DAG->MRI;
+      auto HasExternalAndLocalSources = [&MBB, &MRI](const MachineInstr &MI) {
+        if (!MI.isRegSequence() || !MRI.isSSA())
+          return false;
+        const auto NumExternal =
+            count_if(MI.uses(), [&MBB, &MRI](const MachineOperand &MO) {
+              return MO.isReg() && MO.getReg().isVirtual() &&
+                     MRI.getVRegDef(MO.getReg())->getParent() != &MBB;
+            });
+        const auto NumInternal = MI.getNumOperands() - 1 - (2 * NumExternal);
+        return NumExternal == 1 && NumInternal >= 1;
```

Comment:
nit: `NumExternal > 0`? This at least is better covered by the name of the lambda.

Reply:
See https://github.com/Xilinx/llvm-aie/pull/220#discussion_r1819091008: I'd like to keep the heuristic as is for now